KP-8714 Harvest from COMEDI

aajarven opened the pull request:
Because META-SHARE has been discontinued, we need to switch our metadata source to COMEDI.

TODO:
- [X] Edit metadata parsing & test data to match the CMDI format from COMEDI
- [x] Verify that the parsing makes sense and works for the full data set (e.g. by running parsing only against COMEDI)
- [x] Make the rest of the tests compatible with CMDI/COMEDI
- [x] Verify that staging Metax accepts the outputs
  - NB: there are known issues, e.g. actors without organizations (see [KP-9134](https://jira.eduuni.fi/browse/KP-9134) for details) and these are not expected to work. All records without known issues were accepted into staging Metax on a test run though.

Potential issues:
- [x] In CMDI, there are separate actor elements for persons and organizations, e.g `distributionRightsHolderOrganization`/`distributionRightsHolderPerson `. These are not handled properly at the moment: we might need e.g. to allow using more than one xpath expression for one type of role
- [x] metax_catalog_id is parametrized in config but hard-coded in `harvester/metadata_parser.py`
- [x] it would be nice if our test data didn't have personal information of random people
- [x] license parsing was kind of rushed, so that's one of the more suspicious changes in the PR at the time of writing
- [ ] hard-coding the FIN-CLARIN set for OAI-PMH is a bit nasty (but not likely really a problem, so making that configurable is optional)

aajarven committed:
KP-8714 Fix CLI invocation syntax

aajarven committed:
KP-8714 Metadata parsing for CMDI format from COMEDI

The test data has been replaced with roughly corresponding data from
COMEDI, and the metadata parser has been adjusted so that the correct
fields are extracted from the test data set records. Despite the tests
for metadata parser passing, the code is not always pretty nor is it
even guaranteed to work with a wider selection of records from the full
data set, so further testing is needed.

aajarven committed:
KP-8714 Harvest CMDI format for records in FIN-CLARIN set

Hard-coding the set is not pretty, so that might not be a good long term
solution. It works for testing the code out though.

aajarven committed:
KP-8714 Make tests for `pmh_interface` use CMDI data

aajarven committed:
KP-8714 Update help to match the new file based configuration

aajarven committed:
KP-8714 Fix CLI tests

aajarven committed:
KP-8714 Get rid of Metashare in names and comments

As an added bonus, the test data files that deliver the full OAI-PMH
response are prefixed with `comedi_list_records_` to separate them from
the test files that only contain a single `record`. The former are
needed when testing involves the OAI-PMH requests whereas the latter are
used when testing just the metadata parsing.

aajarven committed:
KP-8714 Remove unused fixture

aajarven committed:
KP-8714 Use the configured `metax_catalog_id` everywhere

Previously the catalog was hard-coded into the metadata parser, but now
everything follows the configuration file.

aajarven committed:
KP-8714 Remove personal data from the test metadata

aajarven committed:
KP-8714 Fix wrong XML element name for distribution rights holder

aajarven committed:
KP-8714 Make publisher parsing work for persons

CMDI has separate distributionRigtsHolderPerson and
distributionRightsHolderOrganization elements, which is a bit tricky for
the current parsing. At the moment, sole organizations are not added as
actors at all though:
```
$ git show b815790f
commit b815790f952b20950564a768e5cf3157c0724cf3
Author: Anni Järvenpää <anni.jarvenpaa@csc.fi>
Date:   Tue Feb 27 11:37:54 2024 +0000

    KP-7427 Skip empty actors

    Because the organizations have not been implemented yet, some of the
    actors had just a role and an empty person dict. Metax does not like
    this, so those are skipped altogether now.
```

Thus it's less bad to read the person element than to read the
organization element. This needs to be made to work with both
organizations and persons though if (when?) we remove the person data
requirement is lifted.

aajarven committed:
KP-8714 Use Finnish organization name if English not available

Previously we thought that English names would be available for all
organizations, but at the time of writing
https://clarino.uib.no/comedi/editor/lb-2024060521 does not have one.

aajarven committed:
KP-8714 Include organization-only actors

On the contrary to the previous implementation, Metax requires each
actor to have at least URL for the organization ("Each actor has an
organization and optionally a person"
https://metax-service.fd-staging.csc.fi/v3/docs/user-guide/datasets-api/#actors).
At this stage, support was added to actors without person info, but
nothing has (yet) been done to the fact that we will still offer Metax
actors without an organization (common e.g. for metadata creators).

aajarven committed:
KP-8714 Report homepage URL to Metax instead of identifier

The current version of Metax API requires URL, so we must provide that.
The data was already read from the URL field in CMDI, do it makes sense
to just ditch identifier and use URL only.

Traubert reviewed (commented):
Tests complete ok and the code changes seem pretty safe. Some consistency questions:

Is the identifier `MSRecordParser` legacy from *M*eta*S*hare? Not a big deal in itself, but there's still one test where it apparently parses `META-XMLSchema`, namely `tests/test_data/missing_pid.xml`, even though its `__init__()` says "Create a CMDI record object".

By rushed license parsing, do you mean `_get_license_url_from_documentation()` etc. in `metadata_parser.py`? I guess there are some slightly fragile things. Could the condition

```
if title_element and "license" in title_element[0].lower():
```

ever have undesired behaviour, for example if there's an element like "licensee"? Or

```
if word.startswith("http://urn.fi")
```

What about https, or no protocol specification at all?

But even these may be ok if there's reason to think that it won't be a problem.

aajarven committed:
KP-8714 Remove Metashare reference from record parser class name

aajarven committed:
KP-8714 Remove unused import

aajarven committed:
KP-8714 Add missing XML namespace

aajarven committed:
KP-8714 Remove handling for missing PIDs as normal situation

All records in Comedi are assigned a PID immediately at creation,
rendering the old Metashare practice of indicating published metadata by
adding the self link void. Comedi indicates the status of a record with
a separate status field ("in-progress" or "published").

This field controls whether the record is available via OAI-PMH (see
[docs](https://clarino.uib.no/comedi/documentation/comedi-documentation#status)):
> Metadata that is being created has status in progress. When your
> metadata is complete and valid, you can change the status to published.
> This means that your metadata is available for hosting via the OAI-PMH
> endpoint.

Thus there's no need to check for PID existence at all. Instead, a
missing PID is clearly an error situation, resulting in an exception
being raised.

aajarven committed:
KP-8714 Make license URN parsing more robust

It is the norm to use http protocol with URNs, so I suppose the old
logic worked fine for all our data, but this is likely a bit more
robust.

aajarven commented:
`MSRecordParser` is legacy: removed the MS part from the name.

The `META-XMLSchema` was a good catch, I didn't think to grep for that. Fixing that also led me to realizing a bigger thing that was previously overlooked. If I recall correctly, during the Metashare era, the process of differentiating between published and in progress records was by adding the self link only when the metadata is complete. In Comedi, all records get assigned a PID immediately, and there's a separate "status" field for this, and it conveniently also controls whether the record is included in OAI-PMH results. Thus missing PIDs are not business as usual any more, and are instead treated as an error.

What I have been thinking about three months ago might be an unsolvable mystery, but my best guess would be that it's about me not having done careful examination whether the currently handled elements are the only places where license information can be found and thus we would need to inspect more elements. Or maybe about possible errors related to the slight refactoring I did on the side.

The `if "license:" in doc_unstruct_text or "license" in doc_unstruct_text:` logic, as well as the `word.startswith("http://urn.fi")` check , were there already, I suppose added in after some discussions between Mietta or Martin and Kaisa. I still changed it to rely on `://urn.fi/urn:nbn:fi` instead, because that is a bit more specific (I don't know if there are other urn.fi urls though) while not assuming a specific protocol.
